EdFi.Ods.Utilities.Migration.Tests/PgSql/MigrationTests/Latest/PostgresGlendaleTests.cs

```diff
@@ -0,0 +1,203 @@
+﻿// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Net;
+using EdFi.Ods.Utilities.Migration.Configuration;
+using EdFi.Ods.Utilities.Migration.Enumerations;
+using EdFi.Ods.Utilities.Migration.Tests.Enumerations;
+using EdFi.Ods.Utilities.Migration.VersionLevel;
+using NUnit.Framework;
+using Shouldly;
+
+namespace EdFi.Ods.Utilities.Migration.Tests.PgSql.MigrationTests.Latest
+{
+    [TestFixture]
+    public abstract class PostgresGlenDaleTests : PostgreSqlMigrationTestBase
+    {
+        protected override DatabaseRestoreSetupOption DatabaseRestoreSetupOption { get; } = DatabaseRestoreSetupOption.RestoreDuringFixtureSetupOnly;
+        protected override string TestDataDirectoryName => "Latest";
+
+        protected OdsUpgradeResult UpgradeResult { get; private set; }
+
+        protected override string OptionalTestSourceOdsBackupFullPath => EnsureTestGlenDaleBackupPathExists();
+
+        protected override string PsqlExecutable => EnsurePsqlExecutablePathExists();
+
+        protected override string TestDisabledReason => string.IsNullOrEmpty(GetFullGlenDaleBackupPath())
+            ? "GlenDale tests are currently optional due to disk space/runtime requirements. A GlenDale backup file was not supplied, so this test will be automatically disabled.  To enable, specify a valid backup file path in the application config"
+            : string.Empty;
+
+        protected string GetGlenDaleBackupDownloadUrl()
+        {
+            var grandBendBackupPathsByVersion = PostgreSqlMigrationTestSettingsProvider.GetGlenDaleBackupPaths();
+            if (grandBendBackupPathsByVersion.ContainsKey(FromVersion))
+            {
+                return grandBendBackupPathsByVersion[FromVersion];
+            }
+
+            throw new ApplicationException($"No GlenDale backup copy Path found for version {FromVersion.DisplayName}.");
+        }
+
+        private static string GetFullGlenDaleBackupPath()
+        {
+            return string.IsNullOrEmpty(PostgreSqlMigrationTestSettingsProvider.GetConfigVariable("GlendaleBackupPath"))
+                ? null
+                : Path.GetFullPath(PostgreSqlMigrationTestSettingsProvider.GetConfigVariable("GlendaleBackupPath"));
+        }
+        
+        protected string EnsureTestGlenDaleBackupPathExists()
+        {
+            var GlenDaleBackupDownloadUrl = GetGlenDaleBackupDownloadUrl();
+            var GlenDaleBackupsDirectory = Path.Combine(TestContext.CurrentContext.TestDirectory, "GlenDaleBackups");
```

Comment:
GlendaleBackups

Reply:
we really need to fix this correctly, and remove this duplication. The base should have abstract methods and the psql download should move up to it.

Reply:
I refactored the code . Please review again